### ISSUE TITLE
Training crash with TE CUDA graphs

### ISSUE THREAD

**Issue by cq-eng** [NVIDIA/Megatron-LM#5966]
# Training crash with TE CUDA graphs (`--cuda-graph-scope attn`) after ~89 iterations on Qwen3-235B

## Environment

| Component | Version |
|---|---|
| Megatron-Core | `0.19.0+ddaa315fe` |
| PyTorch | `2.8.0a0+5228986c39.nv25.6` |
| Transformer Engine | `2.16.0` |

## Configuration

- Qwen3-235B-A22B, TP=2, PP=8, VPP=4, EP=8, ETP=1, DP=16
- BF16
- MoE alltoall dispatcher
- `--overlap-moe-expert-parallel-comm`
- CUDA graph: `--cuda-graph-impl transformer_engine --cuda-graph-scope attn`

## What Happens

Training runs normally for ~80 iterations, then grad norm starts oscillating wildly between normal values and tens of millions, before hitting NaN at iteration 91 and crashing:

| Iteration | Grad Norm |
|---|---|
| 80 | 91 |
| 81 | 463,582 |
| 82 | 50 |
| 83 | 21,210,958 |
| 87 | 127 |
| 88 | 27,622,466 |
| 89 | 11,356,515 |
| 91 | NaN → crash |

The crash message is:

```
ERROR:megatron.core.rerun_state_machine: found NaN in local grad norm for bucket #0
  in backward pass before data-parallel communication collective
```

## Workaround

Removing `--cuda-graph-impl transformer_engine --cuda-graph-scope attn` makes the crash disappear.

## Question

Has anyone seen this before, or knows what might be causing it?

**Comment by yanminjia** [comment]
Hello @buptzyb, @guihong-nv,  it looks a blocker issue in production environment. please let us know if any clue or update. Many thanks.

**Comment by buptzyb** [comment]
CUDA Graph requires bitwise alignment with eager run in deterministic mode. I don't know if there is a silent fail in cudagraph+GDN. Have you checked the first steps after graph capturing? Everything should be matched, including loss, grad norm, num zeros, etc.

**Comment by yanminjia** [comment]
No, the model Qwen3-235B-A22B we trained does not have the Gated DeltaNet (GDN) layers. In some iterations after graph capturing, grad norms reach to 20+ millions. I'm not sure if any kind of memory corruption if cuda graph comes into effective.

**Comment by CodersAcademy006** [comment]
I don't have hardware at this scale to reproduce this, but the failure pattern lines up with a specific gap in the `--cuda-graph-scope attn` + `--overlap-moe-expert-parallel-comm` interaction, so sharing in case it saves debugging time.

With `--cuda-graph-scope attn`, only `CudaGraphModule.attn` is added to `cuda_graph_modules` (`is_graphable_layer` in `cuda_graphs.py`). For an MoE layer that means the router, token dispatch/combine, and experts all run eager. Only self-attention is captured into the graph.

Since TE 2.7 (TE 2.16.0 in this report), Megatron passes `_reuse_graph_input_output_buffers=True` to `make_graphed_callables()` whenever it's available (`is_te_min_version("2.7.0")` check in `cuda_graphs.py`). With that flag, TE weak-refs a layer's static output tensors as soon as its own `_order`-driven schedule marks them as no longer needed, so the caching allocator can hand that same memory to the next graph capture/replay in the interleaved microbatch/chunk order.

That `_order` is built by `get_overlap_moe_expert_parallel_comm_order()`, which only reasons about graphable layers. It has no entry for the eager MoE alltoall dispatch/combine that `overlap_moe_expert_parallel_comm` schedules on a separate comm stream between attn graph replays (`get_comm_stream()` in `pipeline_parallel/utils.py`). So the attn graph's output buffer, which feeds the MoE token dispatch, could be reclaimed and overwritten by the next interleaved attn-graph replay while the async alltoall on the comm stream is still reading it. The outcome would depend on the relative timing of the NCCL alltoall completion versus the next graph replay, which fits the report: healthy for around 80 iterations, then intermittent grad norm corruption before a hard NaN, rather than a failure from step 1.

If this is on the right track, a few things that would help narrow it down:
- Does the crash still happen with `--overlap-moe-expert-parallel-comm` off, same `--cuda-graph-scope attn`? That isolates whether the reorder/overlap path is required to trigger it.
- Does it happen at lower EP/DP, or does it need something close to this scale to expose the timing window?
- Does pulling the router into the graph too (`--cuda-graph-scope attn,moe_router`, dispatch itself still eager) change anything, or does the race live purely in the dispatch/combine step regardless of router scope?

Happy to be wrong here. This is from reading `cuda_graphs.py`, `moe_layer.py`, `token_dispatcher.py`, and TE's `graph.py`, not an actual repro.

**Comment by buptzyb** [comment]
Adding @lhb8125 & @Wohox if we suspect this is about cudagraph + EP overlap.

I remember we usually set `--cuda-graph-scope attn moe_router moe_preprocess` with EP overlap, right? Is attn-only well-tested? @Wohox

**Comment by cq-eng** [comment]
@CodersAcademy006 @buptzyb Thanks for the detailed reply and the helpful ideas. I'll give them a try and let you know how it goes. Appreciate your help!

**Comment by buptzyb** [comment]
@cq-eng Also recommend our cudagraph debug tool. This can help you check and align the tensor value inside graph scope. ([github](https://github.com/buptzyb/torch-cudagraph-debug), [zhihu](https://zhuanlan.zhihu.com/p/2056107363556447866))
